cmd/mt-whisper-importer-writer/main.go

```diff
@@ -0,0 +1,289 @@
+	if err != nil {
+		panic(fmt.Sprintf("Error creating listener: %q", err))
+	}
+}
+
+func throwError(msg string) {
+	msg = fmt.Sprintf("%s\n", msg)
+	if *exitOnError {
+		panic(msg)
+	} else {
+		printLock.Lock()
+		fmt.Fprintln(os.Stderr, msg)
+		printLock.Unlock()
+	}
+}
+
+func log(msg string) {
+	if *verbose {
+		printLock.Lock()
+		fmt.Println(msg)
+		printLock.Unlock()
+	}
+}
+
+func (s *Server) chunksHandler(w http.ResponseWriter, req *http.Request) {
+	metric := &archive.Metric{}
+	err := metric.UnmarshalCompressed(req.Body)
+	if err != nil {
+		throwError(fmt.Sprintf("Error decoding metric stream: %q", err))
+		return
+	}
+	log("Handling new metric")
+
+	if len(metric.Archives) == 0 {
+		throwError("Metric has no archives")
+		return
+	}
+
+	avg := false
+	if whisper.AggregationMethod(metric.AggregationMethod) == whisper.AggregationAverage {
+		avg = true
+	}
+
+	partition, err := s.Partitioner.Partition(&metric.MetricData, int32(*numPartitions))
+	err = s.Index.AddOrUpdate(&metric.MetricData, partition)
+	if err != nil {
+		throwError(fmt.Sprintf("Error updating metric index: %q", err))
+		return
+	}
+
+	for archiveIdx, a := range metric.Archives {
+		archiveTTL := a.SecondsPerPoint * a.Points
+		tableTTL, err := s.selectTableByTTL(archiveTTL)
+		if err != nil {
+			throwError(fmt.Sprintf("Failed to select table for ttl %d in %+v: %q", archiveTTL, s.TTLTables, err))
+			return
+		}
+		entry, ok := s.TTLTables[tableTTL]
+		if !ok {
+			throwError(fmt.Sprintf("Failed to get selected table %d in %+v", tableTTL, s.TTLTables))
+			return
+		}
+		tableName := entry.Table
+
+		if !avg || archiveIdx == 0 || !*fakeAvgAggregates {
+			log(fmt.Sprintf(
+				"inserting %d chunks of archive %d with ttl %d into table %s with ttl %d and key %s",
+				len(a.Chunks), archiveIdx, archiveTTL, tableName, tableTTL, a.RowKey,
+			))
+			s.insertChunks(tableName, a.RowKey, tableTTL, a.Chunks)
+		} else {
+			// averaged archives are a special case because mt doesn't store them as such.
+			// mt reconstructs the averages on the fly from the sum and cnt series, so we need
+			// to generate these two series out of raw averaged data by multiplying each point
+			// with the aggregation span and storing the result as sum, cnt is the aggregation span.
+			type sumCntChunks struct {
+				sum *chunk.Chunk
+				cnt *chunk.Chunk
+			}
+
+			// seconds per point is assumed to be the aggregation span
+			aggSpan := a.SecondsPerPoint
+
+			sumArchive := make([]chunk.IterGen, 0, len(a.Chunks))
+			cntArchive := make([]chunk.IterGen, 0, len(a.Chunks))
+			for _, ig := range a.Chunks {
+				T0 := ig.Ts
+				sum := chunk.New(T0)
+				cnt := chunk.New(T0)
+
+				it, err := ig.Get()
+				if err != nil {
+					throwError(fmt.Sprintf("failed to get iterator from itergen: %q", err))
+					continue
+				}
+
+				for it.Next() {
+					ts, val := it.Values()
+					cnt.Push(ts, float64(aggSpan))
+					sum.Push(ts, val*float64(aggSpan))
```

Comment:
> in the current code the resulting avg series would be correct

yes

> but the numbers which it is generated from are misleading.

not just misleading. they are wrong. if you aggregate together 4 points into a sum, then the sum is the sum of those 4 numbers. and if you aggregate 4 numbers using the count function, then the count is 4.

> an imported archive that's using average will probably not have the sum and cnt series enabled in the MT config, so MT wouldn't even expose that data to the user.

true, but I rather not speculate on that.  I rather keep the option open that somebody imports averaged data and later wants to retrieve the summed aggregation for example; even though there's the caveat it may be inaccurate if the original input data contained some nulls, but that's a caveat we can document with the import procedure. other than that the data will be correct and useful.
You get sum (and cnt) for free if you rollup with averages, so might as well make it available.


> I'm worried about situations where this results in a float, what about extremely weird configs where the raw archive has SecondsPerPoint = 6 and the first aggregate has SecondsPerPoint = 55, that's gonna end up in rounding errors while the current solution would still produce the correct avg series. Although I'm not sure what whisper does in a situation like that

don't worry :) I just dug through the carbon and whisper source code and as you can see, it's enforced that this cannot happen. In fact, this gave me inspiration to add these safeguards to #557 

```
  """ Validates an archiveList.
  An ArchiveList must:
  1. Have at least one archive config. Example: (60, 86400)
  2. No archive may be a duplicate of another.
  3. Higher precision archives' precision must evenly divide all lower precision archives' precision.
  4. Lower precision archives must cover larger time intervals than higher precision archives.
  5. Each archive must have at least enough points to consolidate to the next archive

  Returns True or False
  """
```